### ISSUE TITLE
feature flag `if` not being taken into consideration

### ISSUE THREAD

**Issue by WhyNotHugo** [rust-lang/rust-clippy#8784]
### Summary

Code is here -> https://github.com/librespot-org/librespot/blob/650d41b020691fec6c7d2c90c0f6de91ac8635cf/discovery/src/lib.rs#L103-L126

For this block:

```
        let svc;

        cfg_if! {
            if #[cfg(feature = "with-dns-sd")] {
                svc = dns_sd::DNSService::register(
                    Some(name.as_ref()),
                    "_spotify-connect._tcp",
                    None,
                    None,
                    port,
                    &["VERSION=1.0", "CPath=/"],
                )
                .unwrap();

            } else {
                let responder = libmdns::Responder::spawn(&tokio::runtime::Handle::current())?;
                svc = responder.register(
                    "_spotify-connect._tcp".to_owned(),
                    name,
                    port,
                    &["VERSION=1.0", "CPath=/"],
                )
            }
        };
```

`clippy` reports `unneeded late initalization`. I think it can't "see" the that this feature-based `if` exists due to how the macro works.

### Lint Name

needless_late_init

### Version

```text
rustc 1.60.0 (7737e0b5c 2022-04-04)
binary: rustc
commit-hash: 7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c
commit-date: 2022-04-04
host: x86_64-unknown-linux-gnu
release: 1.60.0
LLVM version: 14.0.0
```

**Comment by Alexendoo** [comment]
Surprising, the spans aren't marked as from expansion and have the root syntax context so I'm not sure how to detect this

The following compiles at least, so I can see why this is linting, but to a human it's wrong

```rust
cfg_if::cfg_if! {
    if #[cfg(feature = "with-dns-sd")] {
        let x = 1;
    } else {
        let x = 2;
    }
};
    
println!("{x}");
```

**Comment by Jarcho** [comment]
For `cfg_if` specifically you don't need to declare anything in advance. This works just fine:
```rust
cfg_if! {
    if #[cfg(feature = "with-dns-sd")] {
        let svc = dns_sd::DNSService::register(
            Some(name.as_ref()),
            "_spotify-connect._tcp",
            None,
            None,
            port,
            &["VERSION=1.0", "CPath=/"],
        )
        .unwrap();
    } else {
        let responder = libmdns::Responder::spawn(&tokio::runtime::Handle::current())?;
        let svc = responder.register(
            "_spotify-connect._tcp".to_owned(),
            name,
            port,
            &["VERSION=1.0", "CPath=/"],
        );
    }
};
```

**Comment by WhyNotHugo** [comment]
Oooh, right... the fact that this is correct is very unintuitive... So I guess clippy is not stricly wrong here :thinking:
